### DISCUSSION TITLE
Rust coding sample

### DISCUSSION THREAD

**Discussion by CountryHands** [XRPLBounties/Proposals#37] [Ideas]
I would like to see more rust integration with the ledger

**Comment by LimpidCrypto** [comment]
## Why rust?
Rust is a very popular programming language in the crypto space. Cryptocurrencies like IOTA or Solana use Rust as their main language. Currently we are missing extensive samples to interact with the XRPL using Rust.

## What are the benefits?
#### Getting started with XRP
Adding code samples makes it easy for new developers to get started with XRP. Developers from other blockchains can switch to the XRPL using the language they are used to programming with.
#### New project opportunities 
Rust is a low-level language, being an alternative to C when your project is dealing with embedded systems.
#### Safety
Rust offers memory safety guarantees (ownership -> alternative to a garbage collector) to prevent memory leaks, buffer overflows, etc. Especially for low-level projects this is a huge deal.

---

## xrpl-rust
### What is xrpl-rust?
With [xrpl-rust](https://github.com/sephynox/xrpl-rust), @sephynox and I are building a `no_std` library to interact with the XRP Ledger. It's based off the well known [xrpl-py](https://github.com/XRPLF/xrpl-py) library. Being selected as one of the winners of the [XRPL Grants](https://xrplgrants.org/grantees?query=xrpl-rust) program in wave 4, helps us building it to be a secure and reliable tool, devs and users can rely on. A quick overview:
#### Features:
##### Wallet:
- Random wallet generation
- Wallet generation from seed
##### Clients (`std` and `no_std`):
- `AsyncWebsocketClient`
- `WebsocketClient`
- `AsyncJsonRpcClient`
- `JsonRpcClient`
##### Models:
- Transactions
- Requests
- Responses
- Ledger objects
- Metadata
##### High-level methods:
- `autofill_and_sign`
- `get_account_balance`
- ...
##### Utility functions:
- Currency conversions
- Transaction parsers
- ...

#### Early (yet to come) amendment support
- AMM
- Sidechains
- Hooks

#### `no_std` compatible
xrpl-rust being fully `no_std` compatible enables developers to use it on embedded systems without an operating system.

#### Security
- utilizing crates like zeroize to prevent sensitive data from being leaked accidentally
- Security audit as a last milestone of the XRPL Grants program

### Current state
Our project is currently at an early state. On the main branch we currently "only" have core functionalities and wallet generation.
Our [milestones](https://github.com/sephynox/xrpl-rust/milestones) are public, so you can see what we're planning to do next.

### Releasing process
We plan to push and do beta-releases on the `main` branch as soon as a milestone is reached on the `develop` branch. Qualifying them to be beta-releases makes it clear that developers still have to expect bugs. We won't consider any release as "stable" until the security audit passed.

**We would love to see Rust code samples on [xrpl.org](xrpl.org) utilizing xrpl-rust. This would also be a great opportunity for us to get the community involved in our project and for them to benefit from it at the same time.**

---

## When to propose new bounties?
We would propose new bounties as soon as a new beta was released. Like, when we have completed the networking milestone we would propose a bounty to add code samples for connecting to a node, etc.

## Qualify samples to be experimental
Since the samples would be build with beta-releases we would like to have the samples marked as "experimental".

---

**Please join the discussion and let us know what you think. We believe there is high demand for interacting with the XRPL using Rust and we hope to get support from this program to expand the XRPL developer community :)**

**Reply by wojake** [reply]
Best to post this as its own Discussions thread and mention it here

**Reply by wojake** [reply]
@intelliot @ckniffen Please consider funding this bounty.
